pandas/core/arrays/base.py

```diff
@@ -406,8 +406,29 @@ def argsort(self, ascending=True, kind='quicksort', *args, **kwargs):
         # 1. _values_for_argsort : construct the values passed to np.argsort
         # 2. argsort : total control over sorting.
         ascending = nv.validate_argsort_with_ascending(ascending, args, kwargs)
-        values = self._values_for_argsort()
-        result = np.argsort(values, kind=kind, **kwargs)
+        values, mask = self._values_for_argsort()
+
+        def permutation(mask):
+            # Return a permutation which maps the indices of the
+            # subarray without nan to the indices of the original array.
+            permu = np.arange(len(mask))
+            nan_loc = np.arange(len(mask))[mask]
+            offset = 0
```

Comment:
we have logic like this elsewhere in the codebase perhaps have a look at pandas/core/indexing.py

Reply:
or pandas/core/sorting.py

Reply:
The sub function permutation is simplified.